### DISCUSSION TITLE
Access master model data from the detail create form

### DISCUSSION THREAD

**Discussion by envas** [wintercms/winter#250] [Q&A] [answered]
I have a model with Orders, OrderItems and Products. On the Order form, I have an OrderItem list with create/update options. In the OrderItem create/update form, I can select from the Products list. So far everything is simple and works well.  

Now I have the following request. The product price, which should be on the OrderItem form automatically updated after a product change, is based on the currency. And this currency is saved in the Orders model, there is one currency for all order items.  I tried to solve it, as usual for such lookups, by using the definition

```
price:
   dependsOn: product
```

and the `filterFields($fields, $context = null)` function. It works fine on the update form where I have the filled OrderItem model. Unfortunately, I don't know how to get the Order model, or at least the order id, on the create form. 

I have a feeling that the problem should be solved somewhere in the controller, which has access to the whole form including Order fields, I can't think of any solution. Any idea?

**Comment by LukeTowers** [comment, marked as answer]
Are you using a relation controller to create the OrderItem, and you're looking to get the Order ID from the OrderItem create form? This should be easy enough by looking directly at the values sent over the AJAX request, although that won't work as well if the Order record itself doesn't exist prior to the OrderItem being created.

**Reply by envas** [reply]
Yes, I am using the relation controller. What you write is right for all AJAX onAnything() requests. But I need the order_id in the **model** function `OrderItem::filterFields()`. Unfortunately in the create form the bound model has no data. I added the order_id field to my OrderItem form hoping it will be populated from the existing order model, but it's not. Obviously, the order_id is added later, maybe onSave(). 

What I am missing here is an AJAX request like `onFormFieldChange($recordId, $field)` so that field lookups can be resolved in the controller and not in the model. Oracle Forms had it back in the early nineties. In my opinion, it is conceptually wrong to work with fields in the model. Fields are visual elements. This does not belong in the model object. But I may have missed something important.

**Reply by LukeTowers** [reply]
You might be able to use `relationExtendManageWidget()` on the controller to set the value ahead of time using something like the following:

```php
    /**
     * Extend the OrderItem relation manage widget to ensure that even new OrderItems have access to the parent Order object
     *
     * @param Backend\Classes\WidgetBase $widget
     * @param string $field
     * @param Winter\Storm\Database\Model $model
     */
    public function relationExtendManageWidget($widget, $field, $model)
    {
        if (!(
                $widget instanceof \Backend\Widgets\Form && 
                $model instanceof \Author\Plugin\Models\Order &&
                $field === 'order_items' &&
                $widget->model instanceof \Author\Plugin\Models\OrderItem
        )) {
            return;
        }

        $widget->model->order_id = $model->id;
    }
```

**Reply by envas** [reply]
Bingo, that's it. I missed these RelationController overrides somehow.

**Reply by LukeTowers** [reply]
@envas feel free to submit a PR to the docs to make it more clear for other people! 😉
